### PR TITLE
Added pattern migration guide to the change log

### PR DESCRIPTION
**Pull request by JackStouffer**
Ping @andralex @wilzbach

### CODE REVIEW

File: changelog/pattern-deprecate.dd [hunk 1]
```diff
@@ -8,3 +8,57 @@ $(MREF std, regex) and $(MREF std, algorithm). They will be removed from
 $(MREF std, string) on May 2018.
 
 If you still need to use these, please see $(LINK2 https://github.com/dlang/undeaD, undeaD).
+
+Here's a simple guide on how to migrate your existing code.
```
Comment (CyberShadow) [inline changelog/pattern-deprecate.dd]: Avoid "Here's X" and adjectives describing the text itself ("simple"); consider: "Examples of the deprecated functions, and their modern replacements:"

File: changelog/pattern-deprecate.dd [hunk 2]
```diff
@@ -8,3 +8,57 @@ $(MREF std, regex) and $(MREF std, algorithm). They will be removed from
...
+import std.string;
+import std.utf;
+
+string s = "\tabc";
+
+// old
+s.munch(whitespace);
+
+// new
+size_t i = s.byCodeUnit.countUntil!(a => !isWhite(a));
```
Comment (CyberShadow) [inline changelog/pattern-deprecate.dd]: Do we not have `skipOver` with one argument + alias? Maybe we should.

File: changelog/pattern-deprecate.dd [hunk 3]
```diff
@@ -8,3 +8,57 @@ $(MREF std, regex) and $(MREF std, algorithm). They will be removed from
...
+-------
+import std.string;
+import std.regex;
+
+// old
+bool flag = inPattern('x', "a-z");
+
+// new
+bool flag = !"x".matchFirst(regex("[a-z]")).empty;
+-------
```
Comment (CyberShadow) [inline changelog/pattern-deprecate.dd]:
```d
// old
if (inPattern('x', "a-z")) { ... }

// new
if ("x".matchFirst(regex("[a-z]"))) { ... }
```

Avoids the awkward bool conversion.

File: changelog/pattern-deprecate.dd [hunk 4]
```diff
@@ -8,3 +8,57 @@ $(MREF std, regex) and $(MREF std, algorithm). They will be removed from
...
+$(REF squeeze, std, string):
+-------
+import std.algorithm;
+import std.string;
+
+// old
+"hello".squeeze;
+
+// new
+"hello".unique;
```
Comment (CyberShadow) [inline changelog/pattern-deprecate.dd]: Fairly sure this should be `uniq`, not `unique`.

File: changelog/pattern-deprecate.dd (hunk 2 above)
Comment (JackStouffer) [inline changelog/pattern-deprecate.dd]: https://issues.dlang.org/show_bug.cgi?id=17525

**Comment by JackStouffer** [comment]
@CyberShadow Fixed

**Comment by CyberShadow** [comment]
Looks nice, though it's a bit weird how it's using expressions without side effects as statements on their own, as if they mutate the string like the first example.

**Comment by JackStouffer** [comment]
@CyberShadow I guess that's just part of understanding range code.

I've added a note just in case.

**Comment by CyberShadow** [comment]
> I guess that's just part of understanding range code.

Changelogs are one part of the website that people unfamiliar with D are likely to check, to see how the language is evolving in the present day. So I think we shouldn't be scarce on details there.

Better than a note, would be to simply assign them to a dummy variable.

File: changelog/pattern-deprecate.dd [hunk 5]
```diff
@@ -8,3 +8,60 @@ $(MREF std, regex) and $(MREF std, algorithm). They will be removed from
 $(MREF std, string) on May 2018.
 
 If you still need to use these, please see $(LINK2 https://github.com/dlang/undeaD, undeaD).
+
+The following are examples of the deprecated functions, and their modern replacements.
+Note that the new code is based on the range model, and therefore does not have
+mutate against the original string, unlike the old code.
```
Comment (CyberShadow) [inline changelog/pattern-deprecate.dd]: This is just wrong. `removechars`, `squeeze` and `munch` all return a string; only `munch` also takes the argument by value and mutates it. `inPattern` doesn't return/mutate a string at all.

I think it's better to remove this node and illustrate the example with dummy variables.

File: changelog/pattern-deprecate.dd (hunk 2 above)
Comment (wilzbach) [inline changelog/pattern-deprecate.dd]: `s = s.find!(a => !isWhite(a));` ?

File: changelog/pattern-deprecate.dd [hunk 6]
```diff
@@ -8,3 +8,60 @@ $(MREF std, regex) and $(MREF std, algorithm). They will be removed from
...
+
+$(REF munch, std, string):
+-------
+import std.algorithm;
```
Comment (wilzbach) [inline changelog/pattern-deprecate.dd]: FYI: There's a new [`RUNNABLE_EXAMPLE` macro](https://github.com/dlang/dlang.org/pull/1709), e.g.:

```
$(RUNNABLE_EXAMPLE
---
// my executable code
---
)
```

**Comment by wilzbach** [comment]
This should be targeting `stable` now. Change the target branch via the GH UI and  locally you can use e.g. `git rebase --onto upstream/stable`

File: changelog/pattern-deprecate.dd (hunk 2 above)
Comment (CyberShadow) [inline changelog/pattern-deprecate.dd]: Nice! Ever shorter: `s = s.find!(not!isWhite);`

Comment (wilzbach) [inline changelog/pattern-deprecate.dd]: Agreed, but now you need to import `std.functional` (overall more key strokes) :cry:

Comment (CyberShadow) [inline changelog/pattern-deprecate.dd]: https://github.com/CyberShadow/AutoFix + https://dump.thecybershadow.net/16b165ae82f44e8b0cb61975ace6ecb6/my-dfix.el = I forgot what writing imports manually is like :)

**Comment by JackStouffer** [comment]
Won't be able to get back to this until Monday. Vacation awaits!

File: changelog/pattern-deprecate.dd (hunk 5 above)
Comment (MartinNowak) [inline changelog/pattern-deprecate.dd]: Done

**Comment by JackStouffer** [comment]
Thanks @MartinNowak for handling this.

**Pull request merged**